client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1165 @@
+	}
+	changeScript, err := txscript.PayToAddrScript(addr)
+	if err != nil {
+		return nil, fmt.Errorf("error creating change script: %v", err)
+	}
+	changeIdx := len(baseTx.TxOut)
+	changeOutput := wire.NewTxOut(1, changeScript)
+	changeOutput.Value = int64(remaining - minFee)
+	isDust := dexdcr.IsDust(changeOutput, dcr.nfo.FeeRate)
+	sigCycles := 0
+	if !isDust {
+		// Add the change output with a recalculated fees
+		size = size + dexdcr.P2PKHOutputSize
+		fee := dcr.nfo.FeeRate * uint64(size)
+		changeOutput.Value = int64(remaining - fee)
+		baseTx.AddTxOut(changeOutput)
+		// Find the best fee rate by closing in on it in a loop.
+		tried := map[uint64]uint8{}
+		for {
+			sigCycles++
+			msgTx, signed, err = dcr.node.SignRawTransaction(baseTx)
+			if err != nil {
+				return nil, fmt.Errorf("signing error: %v", err)
+			}
+			if !signed {
+				return nil, fmt.Errorf("incomplete raw tx signature")
+			}
+			size := msgTx.SerializeSize()
+			reqFee := dcr.nfo.FeeRate * uint64(size)
+			if reqFee > remaining {
+				// I can't imagine a scenario where this condition would be true, but
+				// I'd hate to be wrong.
+				dcr.log.Errorf("reached the impossible place. in = %d, out = %d, reqFee = %d, lastFee = %d",
+					totalIn, totalOut, reqFee, fee)
+				return nil, fmt.Errorf("change error")
+			}
+			if fee == reqFee || (reqFee < fee && tried[reqFee] == 1) {
+				// If we've already tried it, we're on our way back up, so this fee is
+				// likely as good as it gets.
+				break
+			}
+			// We must have some room for improvement
+			tried[fee] = 1
+			fee = reqFee
+			changeOutput.Value = int64(remaining - fee)
+			continue
+		}
+	}
+
+	checkHash := msgTx.TxHash()
+	dcr.log.Debugf("%d signature cycles to converge on fees for tx %s", sigCycles, checkHash)
+	txHash, err := dcr.node.SendRawTransaction(msgTx, false)
+	if err != nil {
+		return nil, err
+	}
+	if *txHash != checkHash {
+		return nil, fmt.Errorf("transaction sent, but received unexpected transaction ID back from RPC server. "+
+			"expected %s, got %s", *txHash, checkHash)
+	}
+	if !isDust {
+		dcr.addChange(txHash, uint32(changeIdx))
+	}
+	return msgTx, nil
+}
+
+// For certain dcrutil.Address types.
+type signatureTyper interface {
+	DSA() dcrec.SignatureType
+}
+
+// createSig creates and returns the serialized raw signature and compressed
+// pubkey for a transaction input signature.
+func (dcr *ExchangeWallet) createSig(tx *wire.MsgTx, idx int, pkScript []byte, addr dcrutil.Address) (sig, pubkey []byte, err error) {
+	sigTyper, ok := addr.(signatureTyper)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid address type")
+	}
+
+	priv, pub, err := dcr.getKeys(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sigType := sigTyper.DSA()
+	switch sigType {
+	case dcrec.STEcdsaSecp256k1:
+		sig, err = txscript.RawTxInSignature(tx, idx, pkScript, txscript.SigHashAll, priv)
+	default:
+		sig, err = txscript.RawTxInSignatureAlt(tx, idx, pkScript, txscript.SigHashAll, priv, sigType)
```

Comment:
This seems to be an older version of txscript.  Now it's just `RawTxInSignature` taking a sig type.

Reply:
The most recent dcrd release is still on v2 and we are using v2 throughout dcrdex. We can switch to v3, but since v2 is used throughout dcrdex, maybe another PR?

Reply:
Yah, another PR is fine for that.  And probably not just yet after all since v3 doesn't have a tag yet.